tests/functional_test.py

```diff
@@ -1,88 +1,167 @@
+    assert len(err.value.data) == 0
 
+
+def test_sign_reject(backend, firmware, navigator, test_name):
     xrp = XRPClient(backend, firmware, navigator)
 
-    prefix = (
-        os.path.dirname(os.path.realpath(__file__)) + f"/snapshots/{firmware.device}/"
-    )
-    full_snappath = pathlib.Path(
-        raw_tx_path.replace("/testcases/", f"/snapshots/{firmware.device}/")
-    ).with_suffix("")
-    no_prefix_snappath = str(full_snappath)[len(prefix) :]
+    # pragma pylint: disable=line-too-long
+    # Transaction extracted from testcases/01-payment/01-basic.raw
+    transaction = "120000228000000024000000036140000000000F424068400000000000000F732102B79DA34F4551CA976B66AA78A55C43707EC2BB2BEC39F95BD53F24E2E45A9E6781140511E17DB83BB6F113939D67BC8EA539EDC926FC83140511E17DB83BB6F113939D67BC8EA539EDC926FC"
+    # pragma pylint: enable=line-too-long
+
+    # Convert message to bytes
+    message = bytes.fromhex(transaction)
 
+    # Send the APDU (Asynchronous)
+    with pytest.raises(ExceptionRAPDU) as err:
+        with xrp.sign(DEFAULT_BIP32_PATH + message):
+            util_navigate(firmware, navigator, test_name, "Reject_sign")
+
+    # Assert we have received a refusal
+    assert err.value.status == Errors.SW_WRONG_ADDRESS
+    assert len(err.value.data) == 0
+
+
+def test_sign_valid_tx(backend, raw_tx_path, firmware, navigator):
     if raw_tx_path.endswith("19-really-stupid-tx.raw"):
         pytest.skip(f"skip invalid tx {raw_tx_path}")
 
+    xrp = XRPClient(backend, firmware, navigator)
+
     with open(raw_tx_path, "rb") as fp:
         tx = fp.read()
 
-    path = Bip32Path.build(DEFAULT_PATH)
-    payload = path + tx
+    index = raw_tx_path.index("/testcases/") + len("/testcases/")
+    snapdir = Path(raw_tx_path[index :]).with_suffix("")
 
     backend.wait_for_home_screen()
-    xrp.sign(payload, True, no_prefix_snappath)
-    assert xrp._client.last_async_response.status == 0x9000
-
-    # Verify tx signature (Does not work...)
-    # key = xrp.get_pubkey()
-    # print(f"RECEIVED PUBKEY : {key}")
-    # print(f"RECEIVED PUBKEY binascii: {binascii.hexlify(key)}")
-    # print(f"RECEIVED PUBKEY binascii str:")
-    # print(str(binascii.hexlify(key), encoding="utf-8"))
-    # xrp.verify_ecdsa_secp256k1(tx, signature, str(binascii.hexlify(b'22'), encoding="utf-8"))
+    with xrp.sign(DEFAULT_BIP32_PATH + tx):
+        util_navigate(firmware, navigator, snapdir, "Sign transaction", False)
+
+    reply = xrp.get_async_response()
+    assert reply.status == Errors.SW_SUCCESS
+
+    # Get Public key
+    key_data, _ = calculate_public_key_and_chaincode(
+        CurveChoice.Secp256k1, DEFAULT_PATH, compress_public_key=True)
+    pub_key = bytearray.fromhex(key_data)
+
+    # Check single/multi signature
+    test_config = Path(raw_tx_path.replace(".raw", ".json"))
+    with open(test_config, "r", encoding="utf-8") as fp:
+        cfg_data = json.load(fp)
+    if "SigningPubKey" in cfg_data and cfg_data["SigningPubKey"] == "":
```

Comment:
Maybe put all this logic in the `verify_ecdsa_secp256k1` function you would just give the tx as parameter and it would compute the hash digest inside.

But it's test code, so it would be nice to have but not a show stopper if it stays like that.

Reply:
Done, and move also the function to `utils` module (makes more sense)

Reply:
Nice ! : )